core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -232,117 +288,171 @@ public boolean contains(IracTombstoneInfo tombstone) {
-                  }
-                  staleTombstones.put(info.primary(), new BackPressure<>(commandsFactory.buildIracTombstonePrimaryCheckCommand(batchSize), rsp));
-               }
-               continue;
-            }
-            stage.dependsOn(new CleanupTask(tombstone).checkRemoteSites());
-         }
+      boolean trace = log.isTraceEnabled();
 
-         //sending any pending stale tombstones
-         for (Map.Entry<Address, BackPressure<IracTombstonePrimaryCheckCommand, Void>> entry : staleTombstones.entrySet()) {
-            BackPressure<IracTombstonePrimaryCheckCommand, Void> backPressure = entry.getValue();
-            if (backPressure.element.isEmpty()) {
-               continue;
-            }
-            if (backPressure.delay == null) {
-               rpcManager.sendTo(entry.getKey(), backPressure.element, DeliverOrder.NONE);
-            } else {
-               backPressure.delay.thenRunAsync(() -> rpcManager.sendTo(entry.getKey(), backPressure.element, DeliverOrder.NONE), blockingExecutor);
-            }
+      if (trace) {
+         log.trace("[IRAC] Starting tombstone cleanup round.");
+      }
 
-         }
-         return stage.freeze().whenComplete(scheduler);
-      } catch (Throwable t) {
-         log.debug("Unexpected exception", t);
-         scheduler.scheduleWithCurrentDelay();
-         return CompletableFutures.completedNull();
+      scheduler.onTaskStarted(tombstoneMap.size());
+
+      CompletionStage<Void> stage = Flowable.fromIterable(tombstoneMap.values())
+            .groupBy(this::classifyTombstone)
+            // We are using flatMap to allow for actions to be done in parallel, but the requests in each action
+            // must be performed sequentially
+            .flatMap(group -> {
+               switch (group.getKey()) {
+                  case REMOVE_TOMBSTONE:
+                     return removeAllTombstones(group);
+                  case NOTIFY_PRIMARY_OWNER:
+                     return notifyPrimaryOwner(group);
+                  case CHECK_REMOTE_SITE:
+                     return checkRemoteSite(group);
+                  case KEEP_TOMBSTONE:
+                  default:
+                     return Flowable.empty();
+               }
+            }, true, ACTION_COUNT, ACTION_COUNT)
+            .lastStage(null);
+      if (trace) {
+         stage = stage.whenComplete(TRACE_ROUND_COMPLETED);
       }
+      return stage.whenComplete(scheduler);
    }
 
    private DistributionInfo getSegmentDistribution(int segment) {
       return distributionManager.getCacheTopology().getSegmentDistribution(segment);
    }
 
-   private final class CleanupTask implements Function<Boolean, CompletionStage<Void>>, Runnable {
-      private final IracTombstoneInfo tombstone;
+   private Flowable<Void> removeAllTombstones(Flowable<IracTombstoneInfo> flowable) {
+      return flowable.concatMapDelayError(tombstone -> {
```

Comment:
I believe it only delays the error that was propagated by the return flowable. You probably want to use a try/catch and return a `Flowable.error`. Otherwise if the removeTombstone method throws an exception it cancels the subscription early.